### PR TITLE
JEP-26 Implement tasks 16-19

### PR DESCRIPTION
**Pull request by skynetplace**

### CODE REVIEW

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 1]
```diff
@@ -216,4 +216,87 @@ public static int[] multiplyArrays(int[] firstIntArray, int[] secondIntArray) {
...
+                    break;
+                }
+            }
+            if (j == chars2.length) {
+                return true;
+            }
+        }
+        return false;
+    }
+
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/TwentyMethods.java:301]: replace this logic using String class

File: src/main/java/com/honcharenko/Starter.java [hunk 2]
```diff
@@ -85,5 +87,33 @@ public static void main(String[] args) {
...
+
+        System.out.println("\n\n18) Принимает 3 инта: размер выходного массива, нижняя граница, верхняя граница. Возвращает массив интов заданой длинный, который содержит случайные числа от нижней границы до верхней границы.");
+        int int_1 = 10;
+        int int_2 = -5;
+        int int_3 = 0;
+
+        int[] randomArray = TwentyMethods.generateRandomArray(int_1, int_2, int_3);
+        System.out.println(Arrays.toString(randomArray));
+
+        System.out.println("\n\n19) Принимает 2 массива чаров, проверяет есть ли в 1 массиве такая же последовательность символов, которую представляет собой второй массив. Возвращает булеан.");
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/Starter.java:112]: add cases:
negative scenario
1st array is empty
2nd array isEmpty

File: src/main/java/com/honcharenko/Starter.java [hunk 3]
```diff
@@ -85,5 +87,33 @@ public static void main(String[] args) {
...
+        int[] intArray = new int[]{12, 34, 81, 6, 7, 1};
+        int[] reversedIntArray = TwentyMethods.reverseIntArray(intArray);
+        for (int item : reversedIntArray) {
+            System.out.println(item);
+        }
+
+        System.out.println("\n\n18) Принимает 3 инта: размер выходного массива, нижняя граница, верхняя граница. Возвращает массив интов заданой длинный, который содержит случайные числа от нижней границы до верхней границы.");
+        int int_1 = 10;
+        int int_2 = -5;
+        int int_3 = 0;
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/Starter.java:107]: int arraySize = 10;
int lowerBound=-5;
int upperBound=0;

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 4]
```diff
@@ -216,4 +216,87 @@ public static int[] multiplyArrays(int[] firstIntArray, int[] secondIntArray) {
...
+            if (!isMatched) {
+                resultArray[index++] = ints2[i];
+            }
+        }
+
+        int[] finalArray = new int[index];
+        System.arraycopy(resultArray, 0, finalArray, 0, index);
+
+        return finalArray;
+    }
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/TwentyMethods.java:258]: try to find another way not to make O(n*n) algorithm complexity

File: src/main/java/com/honcharenko/Starter.java [hunk 5]
```diff
@@ -85,5 +87,33 @@ public static void main(String[] args) {
...
+
+        int[] randomArray = TwentyMethods.generateRandomArray(int_1, int_2, int_3);
+        System.out.println(Arrays.toString(randomArray));
+
+        System.out.println("\n\n19) Принимает 2 массива чаров, проверяет есть ли в 1 массиве такая же последовательность символов, которую представляет собой второй массив. Возвращает булеан.");
+        char[] array1 = {'a', 'b', 'c', 'd', 'e'};
+        char[] array2 = {'c', 'd'};
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/Starter.java:114]: fisstArray
secondArray

File: src/main/java/com/honcharenko/Starter.java [hunk 6]
```diff
@@ -85,5 +87,33 @@ public static void main(String[] args) {
...
+        int[] secondArray = new int[]{45, 77, 3, 21, 1};
+        int[] resultArray = TwentyMethods.findNonMatchingElements(firstArray, secondArray);
+        System.out.println(Arrays.toString(resultArray));
+
+        System.out.println("\n\n17) Принимает масив интов, возвращает его же, но в реверсном порядке.");
+        int[] intArray = new int[]{12, 34, 81, 6, 7, 1};
+        int[] reversedIntArray = TwentyMethods.reverseIntArray(intArray);
+        for (int item : reversedIntArray) {
+            System.out.println(item);
+        }
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/Starter.java:102]: System.out.println(Arrays.toString(reversedIntArray));

File: src/main/java/com/honcharenko/Starter.java [hunk 7]
```diff
@@ -1,5 +1,7 @@
 package com.honcharenko;
 
+import java.util.Arrays;
+
 public class Starter {
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/Starter.java:5]: image
do not oversize

**Review by BogdanParashchak — CHANGES_REQUESTED** [review]

**Review by diamandv — COMMENTED** [review]
can not add something new to this review, except - please, delete unnessesery lines (115, 280)

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 8]
```diff
@@ -216,4 +216,87 @@ public static int[] multiplyArrays(int[] firstIntArray, int[] secondIntArray) {
...
+    /**
+     * 17) Принимает масив интов, возвращает его же, но в реверсном порядке.
+     */
+    public static int[] reverseIntArray(int[] ints) {
+        int[] result = new int[ints.length];
+        for (int i = 0; i < ints.length; i++) {
+            result[i] = ints[ints.length - 1 - i];
+        }
+        return result;
+    }
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java:269]: Done.

File: src/main/java/com/honcharenko/Starter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java:107]: Done.

File: src/main/java/com/honcharenko/Starter.java (hunk 7 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java:5]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 9]
```diff
@@ -216,4 +216,87 @@ public static int[] multiplyArrays(int[] firstIntArray, int[] secondIntArray) {
...
+
+    /**
+     * 18) Принимает 3 инта: размер выходного массива, нижняя граница, верхняя граница. Возвращает массив интов заданой длинный, который содержит случайные числа от нижней границы до верхней границы.
+     */
+    public static int[] generateRandomArray(int size, int lowerBound, int upperBound) {
+        int[] ints = new int[size];
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java:275]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 4 above)
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/TwentyMethods.java:261]: refactor the task:
1. compare index-to-index
2. validate imput for emptiness
3. start with comparing of input arrays lengths

File: src/main/java/com/honcharenko/Starter.java [hunk 10]
```diff
@@ -85,5 +87,33 @@ public static void main(String[] args) {
...
+        System.out.println(Arrays.toString(resultArray));
+
+        System.out.println("\n\n17) Принимает масив интов, возвращает его же, но в реверсном порядке.");
+        int[] intArray = new int[]{12, 34, 81, 6, 7, 1};
+        int[] reversedIntArray = TwentyMethods.reverseIntArray(intArray);
+        for (int item : reversedIntArray) {
+            System.out.println(item);
+        }
+
+        System.out.println("\n\n18) Принимает 3 инта: размер выходного массива, нижняя граница, верхняя граница. Возвращает массив интов заданой длинный, который содержит случайные числа от нижней границы до верхней границы.");
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 11]
```diff
@@ -216,4 +216,87 @@ public static int[] multiplyArrays(int[] firstIntArray, int[] secondIntArray) {
...
+
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 1 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java]: Done.

File: src/main/java/com/honcharenko/Starter.java (hunk 2 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 4 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java]: Done.

File: src/main/java/com/honcharenko/Starter.java (hunk 6 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java]: Done.

File: src/main/java/com/honcharenko/Starter.java (hunk 5 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/TwentyMethods.java]: OK

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 12]
```diff
@@ -250,53 +250,62 @@ public static int[] findNonMatchingElements(int[] ints1, int[] ints2) {
...
-     * 18) Принимает 3 инта: размер выходного массива, нижняя граница, верхняя граница. Возвращает массив интов заданой длинный, который содержит случайные числа от нижней границы до верхней границы.
+     * 18) Принимает 3 инта: размер выходного массива, нижняя граница, верхняя граница.
+     * Возвращает массив интов заданой длинный, который содержит случайные числа от нижней границы до верхней границы.
      */
     public static int[] generateRandomArray(int size, int lowerBound, int upperBound) {
+        //TODO rewrite using exceptions (later)
+        if (size <= 0) {
+            System.out.println("Negative or zero array size!!!");
+            return new int[]{};
+        } else if (lowerBound > upperBound) {
```
Comment (BogdanParashchak) [inline src/main/java/com/honcharenko/TwentyMethods.java:279]: delete this `else`

File: src/main/java/com/honcharenko/Starter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/Starter.java]: OK

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/TwentyMethods.java]: OK

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 4 above)
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/TwentyMethods.java]: I dont see in last commit

File: src/main/java/com/honcharenko/Starter.java (hunk 5 above)
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/Starter.java]: OK

File: src/main/java/com/honcharenko/Starter.java (hunk 6 above)
Comment (BogdanParashchak) [reply src/main/java/com/honcharenko/Starter.java]: OK

File: src/main/java/com/honcharenko/TwentyMethods.java (hunk 4 above)
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java]: Rewrited 16's method.

**Review by BogdanParashchak — APPROVED** [review]

File: src/main/java/com/honcharenko/Starter.java [hunk 13]
```diff
@@ -5,115 +5,371 @@
...
+    }
+
+    private static void test_1_case_2() {
+        char[] chars = {};
+        TwentyMethods.printCharArray(chars);
+    }
+
+    private static void test_1_case_1() {
+        char[] chars = {'e', 'b', '@', '&', '^'};
+        TwentyMethods.printCharArray(chars);
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/Starter.java]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 14]
```diff
@@ -250,53 +250,62 @@ public static int[] findNonMatchingElements(int[] ints1, int[] ints2) {
...
-                    break;
-                }
-            }
-            if (j == chars2.length) {
-                return true;
-            }
+        char[] emptyChars = new char[0];
+        if (Arrays.equals(chars1, emptyChars) || Arrays.equals(chars2, emptyChars)) {
+            System.out.println("Input arrays must not be empty.");
+            return false;
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java]: Done.

File: src/main/java/com/honcharenko/TwentyMethods.java [hunk 15]
```diff
@@ -207,14 +209,14 @@ public static boolean hasDuplicates(byte[] bytes) {
...
      */
     public static int[] multiplyArrays(int[] firstIntArray, int[] secondIntArray) {
         for (int i = 0; i < firstIntArray.length; i++) {
             firstIntArray[i] *= secondIntArray[i];
         }
         return firstIntArray;
-
     }
 
     /**
```
[the remark this answers is not in the document]
Comment (skynetplace) [reply src/main/java/com/honcharenko/TwentyMethods.java:222]: Done

**Pull request merged**